packages/next/build/webpack/plugins/font-stylesheet-gathering-plugin.ts

```diff
@@ -0,0 +1,121 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { NodePath } from 'ast-types/lib/node-path'
+import { visit } from 'next/dist/compiled/recast'
+import { compilation as CompilationType, Compiler } from 'webpack'
+import { namedTypes } from 'ast-types'
+import { RawSource } from 'webpack-sources'
+import {
+  getFontDefinitionFromNetwork,
+  FontManifest,
+} from '../../../next-server/server/font-utils'
+
+interface VisitorMap {
+  [key: string]: (path: NodePath) => void
+}
+
+export default class FontStylesheetGatheringPlugin {
+  compiler?: Compiler
+  gatheredStylesheets: Array<string> = []
+
+  private parserHandler = (
+    factory: CompilationType.NormalModuleFactory
+  ): void => {
+    const JS_TYPES = ['auto', 'esm', 'dynamic']
+    // Do an extra walk per module and add interested visitors to the walk.
+    for (const type of JS_TYPES) {
+      factory.hooks.parser
+        .for('javascript/' + type)
+        .tap(this.constructor.name, (parser) => {
+          var that = this
+          parser.hooks.program.tap(this.constructor.name, (ast: any) => {
+            // We will only optimize fonts from first party code.
+            if (parser?.state?.module?.resource.includes('node_modules')) {
+              return
+            }
+            visit(ast, {
+              visitCallExpression: function (path) {
```

Comment:
I believe you can tap into the callExpression directly instead of manually traversing the AST:
https://github.com/webpack/webpack/blob/master/lib/javascript/JavascriptParser.js#L236-L260

Reply:
I agree. This might also let us remove the logic from both `next/head` as we could just make the link tags inert here itself.
But I could never get this to work hence added `recast` at the time of conformance work.
FWIW, this is the code i tried
```js
parser.hooks.expression.for('__jsx').tap('FontStylesheetGatheringPlugin', expression => {
 console.log(parser.state.module.resource)
})
```

Reply:
Update:https://twitter.com/_prateekbh/status/1280603310211215361
Done 👍🏻
its not the most readable/straightforward thing, but it'll definitely be faster